base/multi.jl

```diff
@@ -2135,14 +2144,145 @@ macro parallel(args...)
     r = loop.args[1].args[2]
     body = loop.args[2]
-    if na==1
-        thecall = :(pfor($(make_pfor_body(var, body)), $(esc(r))))
+    if na == 1
+        thecall = :(foreach(wait, pfor($(make_pfor_body(var, body)), $(esc(r)))))
     else
         thecall = :(preduce($(esc(reducer)), $(make_preduce_body(var, body)), $(esc(r))))
     end
     localize_vars(thecall)
 end
 
+type ParallelAccumulator{T}
+    f::Function
+
+    length::Int
+    pending::Int
+
+    initial::Nullable{T}
+    value::Nullable{T}
+
+    # A function which returns a length value when input the destination pid.
+    # Used to serialize the same object with different length values depending
+    # on the destination pid.
+    destf::Nullable{Function}
+
+    chnl::RemoteChannel
+
+    ParallelAccumulator(f, len) = ParallelAccumulator{T}(f, len, Nullable{T}())
+
+    ParallelAccumulator(f, len, initial::T) =
+                ParallelAccumulator{T}(f, len, Nullable{T}(initial))
+
+    ParallelAccumulator(f, len, initial::Nullable{T}) =
+                ParallelAccumulator{T}(f, len, initial, RemoteChannel(()->Channel{Tuple{Int, T}}(Inf)))
+
+    ParallelAccumulator(f, len, initial, chnl) =
+                ParallelAccumulator{T}(f, len, initial, Nullable{Function}(), chnl)
+
+    ParallelAccumulator(f, len, initial, destf, chnl) =
+                new(f, len, len, initial, initial, destf, chnl)
+end
+
+set_destf(pacc::ParallelAccumulator, f::Function) = (pacc.destf = f; pacc)
+
+function serialize(s::AbstractSerializer, pacc::ParallelAccumulator)
+    serialize_cycle(s, pacc) && return
+    serialize_type(s, typeof(pacc))
+
+    if isnull(pacc.destf)
+        error("Cannot serialize a ParallelAccumulator from a destination node.")
+    end
+
+    len = get(pacc.destf)(worker_id_from_socket(s.io))
+
+    serialize(s, pacc.f)
+    serialize(s, len)
+    serialize(s, pacc.initial)
+    serialize(s, pacc.chnl)
+    nothing
+end
+
+function deserialize(s::AbstractSerializer, t::Type{T}) where T <: ParallelAccumulator
+    f = deserialize(s)
+    len = deserialize(s)
+    initial = deserialize(s)
+    chnl = deserialize(s)
+
+    return T(f, len, initial, chnl)
+end
+
+
+function push!(pacc::ParallelAccumulator, v)
+    if pacc.pending <= 0
+        throw(AssertionError("Reusing a ParallelAccumulator is not allowed. reset(p::ParallelAccumulator)?"))
+    end
+
+    if !isnull(pacc.value)
+        pacc.value = pacc.f(get(pacc.value), v)
+    else
+        pacc.value = pacc.f(v)
+    end
+    pacc.pending -= 1
+
+    if pacc.pending == 0
+        put!(pacc.chnl, (pacc.length, get(pacc.value)))
+    end
+    pacc
+end
+
+function wait(pacc::ParallelAccumulator)
+    while pacc.pending > 0
+        (n, v) = take!(pacc.chnl)
+        pacc.pending -= n
+        if isnull(pacc.value)
+            pacc.value = pacc.f(v)
+        else
+            pacc.value = pacc.f(get(pacc.value), v)
+        end
+    end
+    return get(pacc.value)
+end
+
+function reset(pacc::ParallelAccumulator)
```

Comment:
Another specialization of the existing exported `reset`. I am OK with a new export `reset!` too.